### ISSUE TITLE
Pyright reports `reportUndefinedVariable` on referencing sibling classes as typing

### ISSUE THREAD

**Issue by McDic** [microsoft/pyright#6380]
**Describe the bug**

Pylance reports `reportUndefinedVariable` when you reference sibling class for type hints.

**Code or Screenshots**

```python
import typing
from dataclasses import dataclass


@dataclass
class TextComponent:
    """
    Represent a single text component used in TextJSON.
    """

    class PlainText(typing.TypedDict):
        text: str

    class ScoreboardValueInner(typing.TypedDict):
        name: str
        objective: str

    class ScoreboardValue(typing.TypedDict):
        score: "ScoreboardValueInner"
```

You will get `"ScoreboardValueInner" is not defined Pylance(reportUndefinedVariable)` on the last line.

**VS Code extension or command-line**

I was running Pylance VSCode extension (v2023.11.10) for my Python codes in Python 3.11.6.

Reference: https://github.com/microsoft/pylance-release/issues/5095

**Comment by erictraut** [comment]
This behavior is correct. The symbol `ScoreboardValueInner` is not visible to the sibling classes. That's how Python scoping rules work. You've surrounded `"ScoreboardValueInner"` with quotes, which are normally used for annotations that reference forward-declared symbols. However, in this case, the symbol isn't forward-declared. It's simply not visible from that location. If you remove the quotes and run the code, you'll see that the Python compiler raises an exception and tells you that `ScoreboardValueInner` is not defined.
